components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs

```diff
@@ -522,87 +581,54 @@ public class LegacyJobOrchestrator : IJobOrchestrator, IPublisherConfigServices,
+                                nodesToRemoveSet.Remove(node);
+                            }
+                            else {
+                                updatedNodes.Add(node);
                             }
                         }
-                        existingGroup.Add(entry);
-                    }
-                }
 
-                if (!existingGroup.Any()) {
-                    throw new MethodCallStatusException((int)HttpStatusCode.NotFound, "Endpoint not found.");
+                        entry.OpcNodes = updatedNodes;
+                    }
                 }
 
-                if (!nodeFound) {
-                    throw new MethodCallStatusException((int)HttpStatusCode.NotFound, "Node not found in endpoint.");
+                // Report error if there were entries that we were not able to find.
+                if (nodesToRemoveSet.Count != 0) {
+                    request.OpcNodes = nodesToRemoveSet.ToList();
+                    var entriesNotFoundJson = _jsonSerializer.SerializeToString(request);
+                    throw new MethodCallStatusException(entriesNotFoundJson, (int)HttpStatusCode.NotFound, "Nodes not found");
                 }
 
-                foreach (var entry in existingGroup) {
-                    if (!entry.OpcNodes.Any()) {
-                        _publishedNodesEntrys.Remove(entry);
-                    }
-                }
+                updatedPublishedNodesEntries = updatedPublishedNodesEntries.Where(entry => entry.OpcNodes.Count > 0).ToList();
 
-                var found = false;
-                await _lockJobs.WaitAsync(ct).ConfigureAwait(false);
-                try {
-                    var jobs = _publishedNodesJobConverter.ToWriterGroupJobs(existingGroup, _legacyCliModel);
-                    if (jobs.Any()) {
-                        foreach (var job in jobs) {
-                            var newJob = ToJobProcessingInstructionModel(job);
-                            if (string.IsNullOrEmpty(newJob?.Job?.Id)) {
-                                continue;
-                            }
-                            foreach (var assignedJob in _assignedJobs) {
-                                if (newJob.Job.Id == assignedJob.Value.Job.Id) {
-                                    _assignedJobs[assignedJob.Key] = newJob;
-                                    found = true;
-                                    break;
-                                }
-                            }
-                            if (!found) {
-                                if (_availableJobs.ContainsKey(newJob.Job.Id)) {
-                                    _availableJobs[newJob.Job.Id] = newJob;
-                                    found = true;
-                                }
-                            }
-                        }
-                    }
+                _publishedNodesEntries.Clear();
+                _publishedNodesEntries.AddRange(updatedPublishedNodesEntries);
```

Comment:
>AddRange [](http://example.com/codeflow?start=39&length=8)

probably we should not do this if updatedPublishedNodesEntries is empty.

Reply:
Passing an empty `IEnumerable` collection to `AddRange()` method should not add anything.

Reply:
that's right, but 2 lines below we call refresh jobs with an empty configuration that tries to do jobs conversions lots of memory allocations, locks and other funky stuff for nothing. We can simply clear the jobs here and not attempt a job refresh

Reply:
I've reverted the parts bellow this to the original code that only processes `existingGroups`.

Regarding `updatedPublishedNodesEntries`: it is intended to contain full updated view of `_publishedNodesEntries`, so whether `updatedPublishedNodesEntries` is empty or not `_publishedNodesEntries` should be updated with it. The idea is that we remove nodes only if the request is not going to generate any errors, meaning that the endpoint and **all** nodes are present in our configuration. So if there are nodes that are not present, then we don't change anything. What do you think about this approach to `UnpublishNodesAsync` ? After a bit of testing I can see that this is not exactly what is happening since `_publishedNodesEntries` and `updatedPublishedNodesEntries` share same objects. So when I update `OpcNodes` of an entry in `updatedPublishedNodesEntries` it is also being updated in `_publishedNodesEntries`. I can fix this with either (1) creating full copies of entries for `updatedPublishedNodesEntries` or (2) do one pass over `_publishedNodesEntries` to see if it contains all the nodes (and throw an exception if it doesn't) and then if it does do the second (2) pass to actually remove them. I think that the second (2) approach should be cheaper since it does not create any copies and going over `_publishedNodesEntries` while having a hashset of what is there to remove is relatively fast.

Reply:
Implemented solution (2) with two passes.